modules/javafx.graphics/src/main/native-glass/gtk/glass_key.cpp

```diff
@@ -373,33 +384,18 @@ static Bool isXkbAvailable(Display *display) {
 JNIEXPORT jint JNICALL Java_com_sun_glass_ui_gtk_GtkApplication__1isKeyLocked
   (JNIEnv * env, jobject obj, jint keyCode)
 {
-    Display* display = gdk_x11_display_get_xdisplay(gdk_display_get_default());
-    if (!isXkbAvailable(display)) {
-        return com_sun_glass_events_KeyEvent_KEY_LOCK_UNKNOWN;
-    }
+    GdkKeymap *keymap = gdk_keymap_get_default();
 
-    Atom keyCodeAtom = None;
     switch (keyCode) {
         case com_sun_glass_events_KeyEvent_VK_CAPS_LOCK:
-            keyCodeAtom = XInternAtom(display, "Caps Lock", True);
-            break;
+            return (gdk_keymap_get_caps_lock_state(keymap))
+                ? com_sun_glass_events_KeyEvent_KEY_LOCK_ON
+                : com_sun_glass_events_KeyEvent_KEY_LOCK_OFF;
 
         case com_sun_glass_events_KeyEvent_VK_NUM_LOCK:
-            keyCodeAtom = XInternAtom(display, "Num Lock", True);
-            break;
-    }
-
-    if (keyCodeAtom == None) {
-        return com_sun_glass_events_KeyEvent_KEY_LOCK_UNKNOWN;
-    }
-
-    Bool isLocked = False;
-    if (XkbGetNamedIndicator(display, keyCodeAtom, NULL, &isLocked, NULL, NULL)) {
-        if (isLocked) {
-            return com_sun_glass_events_KeyEvent_KEY_LOCK_ON;
-        } else {
-            return com_sun_glass_events_KeyEvent_KEY_LOCK_OFF;
-        }
+            return (gdk_keymap_get_num_lock_state(keymap))
```

Comment:
Have you tried testing this with a Wayland backend? I was doing some testing and it seemed that unless JavaFX opened a window we didn't connect with the Wayland server so we couldn't get accurate keyboard layout information. It wouldn't surprise me if this also prevented us from getting accurate keyboard state so the manual test for this code (which does not create a window) might fail. Beyond that I agree we should be migrating away from X11 calls if possible.

Reply:
I have attached a `key_release.c` on the JBS bug, it works better on Wayland. On Xorg when you release the CAPS LOCK it will report as still ON until you press another key.

Compile with `gcc -o key_release key_release.c `pkg-config --cflags --libs gtk+-3.0``

Reply:
I reverted back for now since the CAPS LOCK key release behaviour was not ok.